hacspec/src/poly.rs

```diff
@@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
     t
 }
 
-/// Extended euclidean algorithm to compute the inverse of x in yℤ[x]
-#[inline]
-pub fn extended_euclid<T: TRestrictions<T>>(x: &[T], y: &[T], n: T) -> Result<Vec<T>, &'static str> {
-    let (x, y) = normalize(x, y);
-    println!("euclid: {:?} / {:?}", x, y);
-
-    let mut new_t = vec![T::default(); x.len()];
-    new_t[0] = T::from_literal(1);
+// Pad to degree d and reverse coefficients
+fn reverse<T: TRestrictions<T>>(x: &[T], d: usize) -> Vec<T> {
+    let mut out = vec![T::default(); d+1];
+    for i in 0..std::cmp::min(x.len(), d+1) {
+        out[d-i] = x[i]; //XXX: this could probably be written more nicely with iterators?
+    }
+    out 
+}
 
-    let mut new_r = x.clone();
+/// Subtract quotient (bn/x^bd) from (an/x^ad)
+fn quot_sub<T: TRestrictions<T>>(an: &[T], ad: usize, bn: &[T], bd: usize, n: T) -> (Vec<T>, usize){
+    let cd = std::cmp::max(ad, bd);
+    let x = monomial(T::from_literal(1),1);
+    let mut a = an.to_vec();
+    let mut b = bn.to_vec();
+    for _ in 0..cd-ad {
+        a = poly_mul(&a, &x, n);
+    }
+    for _ in 0..cd-bd {
+        b = poly_mul(&b, &x, n);
+    }
+    (poly_sub(&a, &b, n), cd)
+}
 
-    let mut t = vec![T::default(); x.len()];
-    let mut r = y.clone();
+/// Divide a by x assuming a is a multiple of x (shift right by one)
+fn poly_divx<T: TRestrictions<T>>(v: &[T]) -> Vec<T> {
+    let mut out = vec![T::default(); v.len()-1];
+    for i in 0..out.len() {
+        out[i] = v[i+1]; //XXX: this could probably be written more nicely with iterators?
+    }
+    out 
+}
 
-    while !is_zero(&new_r) {
-        println!("{:?} / {:?}", r, new_r);
-        let q = euclid_div(&r, &new_r, n).0;
+/// Iterate division steps in the constant-time polynomial inversion algorithm
+/// See Figure 5.1 from https://eprint.iacr.org/2019/266
+/// Instead of returning M2kx((u,v,q,r)) in last component, only return v
+fn divstepsx<T: TRestrictions<T>>(nn: usize, t: usize, mut delta: i128, fin: &[T], gin: &[T], n: T) -> (i128, Vec<T>, Vec<T>, (Vec<T>, usize)) {
+    debug_assert!(t >= nn);
+    let mut f = fin.to_vec();
+    let mut g = gin.to_vec();
+
+    // Each of u,v,q,r in (f, i) represents quotient f/x^i
+    // u,v,q,r = 1,0,0,1
+    let mut u = (vec![T::from_literal(1); 1], 0); 
+    let mut v = (vec![T::default(); 1], 0);
+    let mut q = (vec![T::default(); 1], 0);
+    let mut r = (vec![T::from_literal(1); 1], 0);
+
+    for i in 0..nn {
+        // Bring u,v,q,r back to fixed precision t
+        u.0 = fixed_truncate(&u.0, t);
+        v.0 = fixed_truncate(&v.0, t);
+        q.0 = fixed_truncate(&q.0, t);
+        r.0 = fixed_truncate(&r.0, t);
+
+        // Decrease precision of f and g in each iteration
+        f = fixed_truncate(&f, nn-i);
+        g = fixed_truncate(&g, nn-i);
+
+        // TODO: make swap constant time
+        if delta > 0  && g[0] != T::default() {
```

Comment:
So what do we need here to make a cswap? `greater_than_bm` and `not_equal_bm` returning all `0 `/`1` of type `T` and `BitAnd`, `BitXor`, `BitOr` should be sufficient I think.
Just making sure we have everything in `Numeric`.